### PR TITLE
bug: 2082395 Make azure baseDomainResoureGroup optional for private cluster

### PR DESCRIPTION
**Pull request by MayXuQQ**
depend on [PR 6049](https://github.com/openshift/installer/pull/6049), do not add the baseDomainResoureGroup for private cluster

### CODE REVIEW

**Comment by MayXuQQ** [comment]
/assign @jianlinliu 
/assign @vrutkovs 
thanks

File: ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh [hunk 1]
```diff
@@ -29,7 +29,6 @@ cat >> "${CONFIG}" << EOF
 baseDomain: ${BASE_DOMAIN}
 platform:
   azure:
-    baseDomainResourceGroupName: os4-common
```
Comment (jianlinliu) [inline ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh:32]: How about just a simple replacement like this:
```
$(if [[ -z "${PUBLISH}" ]] || [[ "${PUBLISH}" == "External" ]]; then echo "baseDomainResourceGroupName: os4-common"; fi)
```

Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh:32]: I did not see any difference after you resolve this.

Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh:32]: Okay, I saw the minor update in the following lines, if you want a dedicated yq patch block to deal with it, I am okay with it too.

File: ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh [hunk 2]
```diff
@@ -60,3 +59,18 @@ EOF
     echo "${OUTBOUND_TYPE} is not supported yet" || exit 1
   fi
 fi
+set +u
```
[the remark this answers is not in the document]
Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh:62]: Another method is, If you do not want to introduce `PUBLISH` env var for this ref, you can use `yq` to get the present setting from install-config.yaml.

File: ci-operator/config/openshift/verification-tests/openshift-verification-tests-master__installer-rehearse-4.10.yaml [hunk 3]
```diff
@@ -53,7 +53,7 @@ tests:
     cluster_profile: azure-qe
     env:
       BASE_DOMAIN: qe.azure.devcluster.openshift.com
-    workflow: cucushift-installer-rehearse-azure-ipi-proxy-techpreview
+    workflow: cucushift-installer-rehearse-azure-ipi-private
```
Comment (jianlinliu) [inline ci-operator/config/openshift/verification-tests/openshift-verification-tests-master__installer-rehearse-4.10.yaml:56]: I do not think 4.10 private install would pass before the fix of 2082395 is backported. But from the job result, it get passed, obviously there is something wrong in the logic.

And the more important thing is even the code logic get fixed in ipi-conf-azure, but it will affect <=4.10 private cluster install.

Comment (MayXuQQ) [reply ci-operator/config/openshift/verification-tests/openshift-verification-tests-master__installer-rehearse-4.10.yaml:56]: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/29967/rehearse-29967-periodic-ci-openshift-verification-tests-master-installer-rehearse-4.10-installer-rehearse-azure/1547532843159130112/artifacts/ci-operator.log
error msg=  54: variable \"azure_base_domain_resource_group_name\" {\nlevel=error\nlevel=error msg=The root module input variable \"azure_base_domain_resource_group_name\" is not\nlevel=error msg=set, and has no default value. Use a -var or -var-file command line argument\nlevel=error msg=to provide a value for this variable.

**Comment by MayXuQQ** [comment]
@jianlinliu how about review , now dev code just support 4.12 (master), I add version check. Thanks

File: ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh [hunk 4]
```diff
@@ -60,3 +59,28 @@ EOF
     echo "${OUTBOUND_TYPE} is not supported yet" || exit 1
   fi
 fi
+
+set +u
+REQUIRED_OCP_VERSION="4.12"
+isOldVersion=true
+if [ -z ${OCP_VERSION} ] || [ "$(printf '%s\n' "${REQUIRED_OCP_VERSION}" "${OCP_VERSION}" | sort --version-sort | head -n1)" = "${REQUIRED_OCP_VERSION}" ]; then
```
Comment (jianlinliu) [inline ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh:66]: Where set the `OCP_VERSION` value?

File: ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh [hunk 5]
```diff
@@ -60,3 +59,28 @@ EOF
     echo "${OUTBOUND_TYPE} is not supported yet" || exit 1
   fi
 fi
+
+set +u
```
Comment (jianlinliu) [inline ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh:63]: suggest to remove the line.

File: ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh [hunk 6]
```diff
@@ -60,3 +59,28 @@ EOF
...
+    cat > "${PATCH}" << EOF
+platform:
+  azure:
+    baseDomainResourceGroupName: os4-common
+EOF
+    /tmp/yq m -x -i "${CONFIG}" "${PATCH}"
+else
+  echo "Omit baseDomainResourceGroupName for priviate cluster"
+fi
+echo "================dump config=============="
```
Comment (jianlinliu) [inline ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh:84]: Pls not dump out the install-config.yaml to avoid security info is exposed.

**Comment by jianlinliu** [comment]
> @jianlinliu how about review , now dev code just support 4.12 (master), I add version check. Thanks

Per https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/29967/rehearse-29967-periodic-ci-openshift-verification-tests-master-installer-rehearse-4.10-installer-rehearse-azure/1549656238923452416 already failed because of this PR change.

File: ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh (hunk 5 above)
Comment (MayXuQQ) [reply ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh:63]: in order to avoid error when has not defined var, I add this. 
PUBLISH and OCP_VERSION maybe not defined.

File: ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh (hunk 4 above)
Comment (MayXuQQ) [reply ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh:66]: some yaml set OCP_VERSION, some use base_ref, I will deal with base_ref when no OCP_VERSION.

File: ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh [hunk 7]
```diff
@@ -60,3 +59,37 @@ EOF
     echo "${OUTBOUND_TYPE} is not supported yet" || exit 1
   fi
 fi
+
+set +u
+
+if [ -z ${OCP_VERSION} ] && [ -n "${base_ref}" ]; then
```
Comment (jianlinliu) [inline ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh:65]: I do not think each workflow in the whole repo always define `OCP_VERSION` or `base_ref`, that totally depends on prow ci job definition.

File: ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-ref.yaml [hunk 8]
```diff
@@ -36,6 +36,13 @@ ref:
   - name: OUTBOUND_TYPE
     default:  ""
     documentation: |-
-      This value represents the routing strategy used to connect the cluster to the internet 
+      This value represents the routing strategy used to connect the cluster to the internet
+  - name: PUBLISH
```
Comment (jianlinliu) [inline ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-ref.yaml:40]: Let us read `publish` from install-config.yaml instead of getting it from ENV, just like https://github.com/openshift/release/pull/30855/files#diff-56bc3e48dda0b0233a4d83f96ae56773ff74ecd6ed86e316b880d041a92af2e2R26

File: ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh [hunk 9]
```diff
@@ -4,6 +4,28 @@ set -o nounset
 set -o errexit
 set -o pipefail
 
+function getVersion() {
+  release_image=""
+  if [ -z "${RELEASE_IMAGE_INITIAL-}" ]; then
```
Comment (jianlinliu) [inline ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh:9]: Sound like the logic would be like the following and easy reading:
```
if [ -n "${RELEASE_IMAGE_INITIAL-}" ]; then
  release_image="${RELEASE_IMAGE_INITIAL}"
elif  [ -n "${RELEASE_IMAGE_LATEST-}" ]; then
  release_image="${RELEASE_IMAGE_LATEST-}"
fi
```

File: ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh [hunk 10]
```diff
@@ -4,6 +4,28 @@ set -o nounset
 set -o errexit
 set -o pipefail
 
+function getVersion() {
+  release_image=""
```
Comment (jianlinliu) [inline ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh:8]: pls add `local` for the env definition in function.

File: ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh [hunk 11]
```diff
@@ -60,3 +81,26 @@ EOF
     echo "${OUTBOUND_TYPE} is not supported yet" || exit 1
   fi
 fi
+
+version=$(getVersion)
```
Comment (jianlinliu) [inline ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh:85]: If you want to use `getVersion` to receive version value, `getVersion` should not print out any other things, only version value is required.

File: ci-operator/config/openshift/verification-tests/openshift-verification-tests-master__installer-rehearse-4.12.yaml [hunk 12]
```diff
@@ -0,0 +1,47 @@
+base_images:
```
Comment (jianlinliu) [inline ci-operator/config/openshift/verification-tests/openshift-verification-tests-master__installer-rehearse-4.12.yaml:1]: Why this file are changed? You did no introduce any new test item, should have no such new job.

File: ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh [hunk 13]
```diff
@@ -60,3 +75,27 @@ EOF
...
+echo "get ocp version:" ${version}
```
Comment (jianlinliu) [inline ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh:80]: You can move the `${version}` inside `""`

File: ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh [hunk 14]
```diff
@@ -60,3 +75,27 @@ EOF
...
+REQUIRED_OCP_VERSION="4.12"
+isOldVersion=true
+if [ -n "${version}" ] && [ "$(printf '%s\n' "${REQUIRED_OCP_VERSION}" "${version}" | sort --version-sort | head -n1)" = "${REQUIRED_OCP_VERSION}" ]; then
+  isOldVersion=false
+fi
+
+PUBLISH=$(/tmp/yq r "${CONFIG}" "publish")
+echo "publish:" "${PUBLISH}"
```
Comment (jianlinliu) [inline ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh:88]: You can move the ${PUBLISH} and ${isOldVersion} inside ""

File: ci-operator/config/openshift/verification-tests/openshift-verification-tests-master__installer-rehearse-4.12.yaml (hunk 12 above)
Comment (MayXuQQ) [reply ci-operator/config/openshift/verification-tests/openshift-verification-tests-master__installer-rehearse-4.12.yaml:1]: need not test the 4.12 nightly build ?

Comment (jianlinliu) [reply ci-operator/config/openshift/verification-tests/openshift-verification-tests-master__installer-rehearse-4.12.yaml:1]: okay, if you want to test your PR works well in 4.12, you can add this testing.

File: core-services/prow/02_config/_boskos.yaml [hunk 15]
```diff
@@ -608,11 +608,36 @@ resources:
   state: free
   type: azure-2-quota-slice
 - names:
+  - centralus--azure-qe-quota-slice-0
```
Comment (jianlinliu) [inline core-services/prow/02_config/_boskos.yaml:611]: It is better to move adding slice quota in another PR.

**Comment by MayXuQQ** [comment]
checked the failed jobs: 
1. "azure-qe-quota-slice"
2. https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/29967/rehearse-29967-periodic-ci-kata-containers-kata-containers-main-4.11-e2e-tests/1553943572481314816/artifacts/e2e-tests/ipi-install-install/finished.json  success
3. https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/29967/rehearse-29967-periodic-ci-openshift-hypershift-main-periodics-e2e-conformance-azure/1553991130708709376/artifacts/e2e-conformance-azure/ipi-install-install/finished.json  success

File: core-services/prow/02_config/_boskos.yaml [hunk 16]
```diff
@@ -978,6 +978,10 @@ resources:
   min-count: 5
   state: free
   type: openstack-nfv-quota-slice
+- max-count: 2
```
Comment (jianlinliu) [inline core-services/prow/02_config/_boskos.yaml:981]: Sounds relevant change with your PR.

File: ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh [hunk 17]
```diff
@@ -60,3 +75,27 @@ EOF
...
+  echo "Write the 'baseDomainResourceGroupName: os4-common' to install-config"
+  PATCH="${SHARED_DIR}/install-config-baseDomainRG.yaml.patch"
+    cat > "${PATCH}" << EOF
+platform:
+  azure:
+    baseDomainResourceGroupName: os4-common
+EOF
+    /tmp/yq m -x -i "${CONFIG}" "${PATCH}"
+else
+  echo "Omit baseDomainResourceGroupName for priviate cluster"
```
Comment (vrutkovs) [inline ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-commands.sh:100]:
```suggestion
  echo "Omit baseDomainResourceGroupName for private cluster"
```

File: ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-ref.yaml [hunk 18]
```diff
@@ -1,9 +1,9 @@
 ref:
   as: ipi-conf-azure
   from_image:
-   namespace: origin
-   name: centos
-   tag: '8'
+   namespace: ocp
+   name: "4.5"
+   tag: upi-installer
```
Comment (vrutkovs) [inline ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-ref.yaml:6]: Not sure why we want `upi-installer` - and from 4.5 specifically

**Review by vrutkovs — CHANGES_REQUESTED** [review]

File: ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-ref.yaml (hunk 18 above)
Comment (MayXuQQ) [reply ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-ref.yaml:6]: this image contains oc , which is used to get version from the image version "oc adm release info ${release_image}

Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-ref.yaml:6]: @MayXuQQ I guess @vrutkovs was suggesting like this:
```
ref:
  as: ipi-conf-azure
  from: upi-installer
  cli: latest
```

Comment (vrutkovs) [reply ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-ref.yaml]: Lets use `from: cli` here, if possible. 

I don't think we ship `upi-installer` in the payload, so when its truly needed we have to import upi-installer from 4.5 imagesteram, but its not the case here imo

Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-ref.yaml]: @vrutkovs thanks for the info. 

When tried `upi-installer` executor images, the following job have different behaviors. 

https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/29967/rehearse-29967-periodic-ci-openshift-hypershift-main-periodics-e2e-conformance-azure/1554722988799037440  failed with reason ImagePullBackOff and message Back-off pulling image "image-registry.openshift-image-registry.svc:5000/ci-op-p4gxyn88/stable:upi-installer"

https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/29967/rehearse-29967-periodic-ci-openshift-verification-tests-master-installer-rehearse-4.12-installer-rehearse-azure/1554662529622347776 get passed, did not hit such problem, do you know what caused the different results.

Comment (vrutkovs) [reply ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-ref.yaml]: First one is invoked from nightly release - `Resolved release latest to registry.ci.openshift.org/ocp/release:4.11.0-0.ci-2022-08-02-023116` - which doesn't have `upi-installer` included as its not part of the payload.

Second one is invoked for installer, so every installer image - installer, baremetal-installer, upi-installer - are being built and tagged into `stable`. We should avoid using `upi-installer` in periodics as it may not exist in the test payload

Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-ref.yaml]: Got it, thanks a lot.

Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-ref.yaml]: @vrutkovs per the latest test result, seem like cli image does not install `jq`, so maybe we still go back to upi-install:4.5

Comment (vrutkovs) [reply ci-operator/step-registry/ipi/conf/azure/ipi-conf-azure-ref.yaml]: You can import `cli-jq`:
```
  cli-jq:
    name: cli-jq
    namespace: ocp
    tag: latest
```

**Comment by MayXuQQ** [comment]
checked the ipi-install-install of the failed jobs , "passed":true,"result":"SUCCESS"

**Review by vrutkovs — APPROVED** [review]
/approve

**Comment by MayXuQQ** [comment]
- [e2e-conformance-azure](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/29967/rehearse-29967-periodic-ci-openshift-hypershift-main-periodics-e2e-conformance-azure/1555124142347390976/artifacts/e2e-conformance-azure/ipi-install-install/finished.json)

- [azure-e2e-ccm-install](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/29967/rehearse-29967-pull-ci-openshift-windows-machine-config-operator-master-azure-e2e-ccm-install/1555124140791304192/artifacts/azure-e2e-ccm-install/ipi-install-install/finished.json)

- private-release-4.11-nightly-4.11-upgrade-from-stable-4.10-azure-upi-disconnected — Job need be dropped.

**Comment by MayXuQQ** [comment]
/assign @jianlinliu 
/assign @JianLi-RH 
thanks , ci/prow/ordered-prow-config — Job failed. after make jobs
just got the diff
-  cluster: build01
+  cluster: build05
